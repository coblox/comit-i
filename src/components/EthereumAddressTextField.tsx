import { InputAdornment, TextField } from "@material-ui/core";
import IconButton from "@material-ui/core/IconButton/IconButton";
import { StandardTextFieldProps } from "@material-ui/core/TextField";
import AutorenewIcon from "@material-ui/icons/Autorenew";
import React from "react";
import Web3Loader from "./Web3Loader";

interface Props extends StandardTextFieldProps {
  onAddress: (address: string) => void;
}

const styles= {
  minWidth: "27rem"
};

function EthereumAddressTextField({
  className,
  onAddress,
  InputProps,
  ...remainingProps
}: Props) {
  return (
    <TextField
      {...remainingProps}
      style={styles}
      label="Ethereum Address"
      InputProps={{
        ...InputProps,
        endAdornment: (
          <InputAdornment position="end">
            <Web3Loader
              ifPresent={web3 => (
                <IconButton
                  aria-label="Load"
                  onClick={async () => {
                    const accounts = await web3.eth.getAccounts();
                    onAddress(accounts[0]);
                  }}
                >
                  <AutorenewIcon />
                </IconButton>
              )}
            />
          </InputAdornment>
        )
      }}
    />
  );
}

export default EthereumAddressTextField;
