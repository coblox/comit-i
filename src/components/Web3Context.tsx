import React, { ReactNode, useContext } from "react";
import Web3 from "web3";

const Web3Context = React.createContext<Web3 | undefined>(undefined);

interface Web3ProviderProps {
  children: ReactNode;
}

function newWeb3() {
  if (window.web3) {
    const web3 = new Web3(window.web3.currentProvider);

    // Why? Because: https://github.com/ethereum/web3.js/issues/2822
    web3.eth.transactionConfirmationBlocks = 1;

    return web3;
  } else {
    return undefined;
  }
}

export function Web3Provider({ children }: Web3ProviderProps) {
  return (
    <Web3Context.Provider value={newWeb3()}>{children}</Web3Context.Provider>
  );
}

interface Web3LoaderProps {
  ifPresent?: (web3: Web3) => JSX.Element;
  ifNotPresent?: () => JSX.Element;
}

export function Web3Loader({
  ifPresent = () => <span />,
  ifNotPresent = () => <span />
}: Web3LoaderProps) {
  return (
    <Web3Context.Consumer>
      {web3 => (web3 ? ifPresent(web3) : ifNotPresent())}
    </Web3Context.Consumer>
  );
}

export function useWeb3() {
  return useContext(Web3Context);
}
