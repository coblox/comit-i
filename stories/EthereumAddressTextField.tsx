import { storiesOf } from '@storybook/react';
import React from 'react';
import EthereumAddressTextField from "../src/components/EthereumAddressTextField";

storiesOf('EthereumAddressTextField', module)
  .add('default', () => (
    <EthereumAddressTextField/>
  ))
;
