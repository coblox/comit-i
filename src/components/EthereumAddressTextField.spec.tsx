import { ThemeProvider } from "@material-ui/styles";
import React from "react";
import renderer from "react-test-renderer";
import appTheme from "../theme";
import EthereumAddressTextField from "./EthereumAddressTextField";
import { Web3Provider } from "./Web3Context";

describe("EthereumAddressTextField", () => {
  it("should render without Web3 being available", () => {
    const tree = renderer
      .create(
        <ThemeProvider theme={appTheme}>
          <EthereumAddressTextField onAddress={() => undefined} />
        </ThemeProvider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("should render with meta mask icon if Web3 is available", () => {
    const fakeWindow = {
      web3: {
        currentProvider: "http://localhost:8454"
      }
    };

    const tree = renderer
      .create(
        <Web3Provider window={fakeWindow as any}>
          <ThemeProvider theme={appTheme}>
            <EthereumAddressTextField onAddress={() => undefined} />
          </ThemeProvider>
        </Web3Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
