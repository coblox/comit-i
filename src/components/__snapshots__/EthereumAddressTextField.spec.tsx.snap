// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthereumAddressTextField should render with meta mask icon if Web3 is available 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root makeStyles-textField-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
    >
      <button
        aria-describedby={null}
        aria-label="Load"
        className="MuiButtonBase-root MuiIconButton-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        title="Please unlock metamask first."
        type="button"
      >
        <span
          className="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            className="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 318.6 318.6"
          >
            <polygon
              className="makeStyles-st0-60"
              points="274.1,35.5 174.6,109.4 193,65.8 "
            />
            <g>
              <polygon
                className="makeStyles-st1-61"
                points="44.4,35.5 143.1,110.1 125.6,65.8"
              />
              <polygon
                className="makeStyles-st1-61"
                points="238.3,206.8 211.8,247.4 268.5,263 284.8,207.7"
              />
              <polygon
                className="makeStyles-st1-61"
                points="33.9,207.7 50.1,263 106.8,247.4 80.3,206.8"
              />
              <polygon
                className="makeStyles-st1-61"
                points="103.6,138.2 87.8,162.1 144.1,164.6 142.1,104.1"
              />
              <polygon
                className="makeStyles-st1-61"
                points="214.9,138.2 175.9,103.4 174.6,164.6 230.8,162.1"
              />
              <polygon
                className="makeStyles-st1-61"
                points="106.8,247.4 140.6,230.9 111.4,208.1"
              />
              <polygon
                className="makeStyles-st1-61"
                points="177.9,230.9 211.8,247.4 207.1,208.1"
              />
            </g>
            <g>
              <polygon
                className="makeStyles-st2-62"
                points="211.8,247.4 177.9,230.9 180.6,253 180.3,262.3"
              />
              <polygon
                className="makeStyles-st2-62"
                points="106.8,247.4 138.3,262.3 138.1,253 140.6,230.9"
              />
            </g>
            <polygon
              className="makeStyles-st3-63"
              points="138.8,193.5 110.6,185.2 130.5,176.1 "
            />
            <polygon
              className="makeStyles-st3-63"
              points="179.7,193.5 188,176.1 208,185.2 "
            />
            <g>
              <polygon
                className="makeStyles-st4-64"
                points="106.8,247.4 111.6,206.8 80.3,207.7"
              />
              <polygon
                className="makeStyles-st4-64"
                points="207,206.8 211.8,247.4 238.3,207.7"
              />
              <polygon
                className="makeStyles-st4-64"
                points="230.8,162.1 174.6,164.6 179.8,193.5 188.1,176.1 208.1,185.2"
              />
              <polygon
                className="makeStyles-st4-64"
                points="110.6,185.2 130.6,176.1 138.8,193.5 144.1,164.6 87.8,162.1"
              />
            </g>
            <g>
              <polygon
                className="makeStyles-st5-65"
                points="87.8,162.1 111.4,208.1 110.6,185.2"
              />
              <polygon
                className="makeStyles-st5-65"
                points="208.1,185.2 207.1,208.1 230.8,162.1"
              />
              <polygon
                className="makeStyles-st5-65"
                points="144.1,164.6 138.8,193.5 145.4,227.6 146.9,182.7"
              />
              <polygon
                className="makeStyles-st5-65"
                points="174.6,164.6 171.9,182.6 173.1,227.6 179.8,193.5"
              />
            </g>
            <polygon
              className="makeStyles-st6-66"
              points="179.8,193.5 173.1,227.6 177.9,230.9 207.1,208.1 208.1,185.2 "
            />
            <polygon
              className="makeStyles-st6-66"
              points="110.6,185.2 111.4,208.1 140.6,230.9 145.4,227.6 138.8,193.5 "
            />
            <polygon
              className="makeStyles-st7-67"
              points="180.3,262.3 180.6,253 178.1,250.8 140.4,250.8 138.1,253 138.3,262.3 106.8,247.4 117.8,256.4 140.1,271.9 178.4,271.9 200.8,256.4 211.8,247.4 "
            />
            <polygon
              className="makeStyles-st8-68"
              points="177.9,230.9 173.1,227.6 145.4,227.6 140.6,230.9 138.1,253 140.4,250.8 178.1,250.8 180.6,253 "
            />
            <g>
              <polygon
                className="makeStyles-st9-69"
                points="278.3,114.2 286.8,73.4 274.1,35.5 177.9,106.9 214.9,138.2 267.2,153.5 278.8,140 273.8,136.4 281.8,129.1 275.6,124.3 283.6,118.2"
              />
              <polygon
                className="makeStyles-st9-69"
                points="31.8,73.4 40.3,114.2 34.9,118.2 42.9,124.3 36.8,129.1 44.8,136.4 39.8,140 51.3,153.5 103.6,138.2 140.6,106.9 44.4,35.5"
              />
            </g>
            <polygon
              className="makeStyles-st6-66"
              points="267.2,153.5 214.9,138.2 230.8,162.1 207.1,208.1 238.3,207.7 284.8,207.7 "
            />
            <polygon
              className="makeStyles-st6-66"
              points="103.6,138.2 51.3,153.5 33.9,207.7 80.3,207.7 111.4,208.1 87.8,162.1 "
            />
            <polygon
              className="makeStyles-st6-66"
              points="174.6,164.6 177.9,106.9 193.1,65.8 125.6,65.8 140.6,106.9 144.1,164.6 145.3,182.8 145.4,227.6 173.1,227.6 173.3,182.8 "
            />
          </svg>
        </span>
        <span
          className="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</div>
`;

exports[`EthereumAddressTextField should render without Web3 being available 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root makeStyles-textField-1"
>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl MuiInputBase-adornedEnd"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      className="MuiInputBase-input MuiInput-input MuiInputBase-inputAdornedEnd"
      disabled={false}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      required={false}
      type="text"
    />
    <div
      className="MuiInputAdornment-root MuiInputAdornment-positionEnd"
    >
      <span />
    </div>
  </div>
</div>
`;
